File: java/Refactor1/src/main/java/no/bouvet/solid/srpdip/domain/orderoperations/CreateOrderOperation.java
package no.bouvet.solid.srpdip.domain.orderoperations;

import no.bouvet.solid.srpdip.InventoryRepository;
import no.bouvet.solid.srpdip.OrderRepository;
import no.bouvet.solid.srpdip.domain.InventoryItem;
import no.bouvet.solid.srpdip.domain.Order;
import no.bouvet.solid.srpdip.domain.OrderItemState;
import no.bouvet.solid.srpdip.domain.OrderState;
import no.bouvet.solid.srpdip.domain.revenue.PriceCalculator;
import no.bouvet.solid.srpdip.messageinterface.RequestMessage;
import no.bouvet.solid.srpdip.messageinterface.ResponseMessage;
import no.bouvet.solid.srpdip.messageinterface.ResponseMessageFactory;

public class CreateOrderOperation implements OrderOperation {

	private InventoryRepository inventoryRepository;
	private OrderRepository orderRepository;

	private ResponseMessageFactory responseMessageFactory = new ResponseMessageFactory();

	private PriceCalculator priceCalculator = new PriceCalculator();

	public CreateOrderOperation(InventoryRepository inventory, OrderRepository orders) {
		this.inventoryRepository = inventory;
		this.orderRepository = orders;
	}

	@Override
	public ResponseMessage execute(RequestMessage request) {
		Order order = orderRepository.createOrder();

		order.getOrderItems().addAll(request.getOrderItems());

		order.getOrderItems().forEach(item -> {
			InventoryItem inventoryItem = inventoryRepository.getInventoryItem(item.getItemCode());

			if (inventoryItem.getQuantityOnHand() >= item.getQuantity()) {
				inventoryItem.setQuantityOnHand(inventoryItem.getQuantityOnHand() - item.getQuantity());
				item.setWeight(item.getWeightPerUnit() * (float) item.getQuantity());

				priceCalculator.calculate(item, inventoryItem);

				item.setState(OrderItemState.FILLED);
			} else {
				item.setState(OrderItemState.NOT_ENOUGH_QUANTITY_ON_HAND);
			}
		});

		order.setState(order.getOrderItems().stream().allMatch(o -> o.getState() == OrderItemState.FILLED) ? OrderState.FILLED
				: OrderState.PROCESSING);

		orderRepository.addOrder(order);

		// save inventory
		inventoryRepository.updateInventory();

		// save order
		orderRepository.updateOrders();

		return responseMessageFactory.createOrderSubmissionResponseMessage(request, order);
	}

}

File: cs/Refactor2/SRP_DI_Workshop/Domain/OrderOperations/CreateOrderOperation.cs
using System;
using System.Linq;
using SRP_DI_Workshop.Domain.Revenue;
using SRP_DI_Workshop.ServiceInterface;

namespace SRP_DI_Workshop.Domain.OrderOperations
{
    public class CreateOrderOperation : IOrderOperation
    {
        private readonly IPriceCalculator _priceCalculator;
        private readonly IInventoryRepository _inventoryService;
        private readonly IOrderRepository _orderRepository;
        private readonly ILoggerService _loggerService;
        private readonly IResponseMessageFactory _responseMessageFactory;

        public CreateOrderOperation(
            IPriceCalculator priceCalculator, 
            IInventoryRepository inventoryService, 
            IOrderRepository orderRepository, 
            ILoggerService loggerService, 
            IResponseMessageFactory responseMessageFactory)
        {
            _priceCalculator = priceCalculator;
            _inventoryService = inventoryService;
            _orderRepository = orderRepository;
            _loggerService = loggerService;
            _responseMessageFactory = responseMessageFactory;

            Operation = Operation.SubmitOrder;
        }

        public Operation Operation { get; private set; }

        public ResponseMessage ExecuteOperation(RequestMessage request)
        {
            try
            {
                Order order = _orderRepository.CreateOrder();

                order.OrderItems.AddRange(request.OrderItems);

                foreach (OrderItem item in order.OrderItems)
                {
                    InventoryItem inventoryItem = _inventoryService.GetInventoryItem(item.ItemCode);

                    if (item.Quantity <= inventoryItem.QuantityOnHand)
                    {
                        inventoryItem.QuantityOnHand -= item.Quantity;
                        item.Weight = item.WeightPerUnit * (float)item.Quantity;

                        _priceCalculator.CalculatePrice(item, inventoryItem);

                        item.State = OrderItemState.Filled;
                    }
                    else
                    {
                        item.State = OrderItemState.NotEnoughQuantityOnHand;
                    }
                }

                order.State = order.OrderItems.All(o => o.State == OrderItemState.Filled)
                    ? OrderState.Filled
                    : OrderState.Processing;

                _orderRepository.AddOrder(order);

                // save inventory
                _inventoryService.UpdateInventory();

                // save order
                _orderRepository.UpdateOrders();

                return _responseMessageFactory.CreateOrderSubmissionResponseMessage(request, order);
            }
            catch (Exception ex)
            {
                _loggerService.WriteLine("Exception during operation SubmitOrder: " + ex, "SubmitOrderError");
                return _responseMessageFactory.CreateErrorResponseMessage(ex);
            }
        }
    }
}